google/cloud/options.cc

```diff
@@ -40,6 +41,24 @@ Options MergeOptions(Options a, Options b) {
   return a;
 }
 
+namespace {
+
+// The prevailing options for the current operation.  Thread local, so
+// additional propagation must be done whenever work for the operation
+// is done in another thread.
+thread_local Options current_options;
```

Comment:
This obviously has a non-trivial destructor.  I am certain you know better than I do whether this is accepted by the Google Style Guide.  Just asking to maybe learn something.

Reply:
The non-trivial destructor for the namespace-scoped object is not a problem because the object is only accessible to its thread, and is not destroyed until the thread terminates.  So, there is no issue with destruction order, and indeed the Style Guide says as much.

What the Style Guide still does require for namespace-scoped objects, thread local or not, is a compile-time-constant initializer.  In the thread-local case I guess this is to guard against threads being created before global initialization is complete.  The workaround is the same in both cases ... to convert the namespace-scoped object to a function local, so that it is initialized on the first call, no matter when that occurs.  I've done that.  PTAL.

Reply:
Ack, thanks for looking into it.